## Reto: Semana 2
En diciembre de 2015, la Candidatura d'Unitat Popular (CUP) un partido político independentista de izquierda que actúa principalmente en Cataluña-, celebró una votación interna sobre si apoyar o oponerse a la investidura de Artur Mas como presidente de la Generalitat. El resultado fue un empate con 1 515 votos a favor y 1 515 en contra, generando gran debate por la naturaleza del resultado.



Tomando en cuenta esta información:  
**a) Calcula cuál es la probabilidad de este empate bajo un enfoque frecuentista:**  
    * Asumiendo que la probabilidad de votar a favor de Arturo Mas es de 0.5   
    * La decisión de voto es independiente entre los votantes.  
**b) Calcula la probabilidad de empate bajo un enfoque Bayesiano**  
    * Asumiendo que no tenemos conocimiento previo del valor de p = probabilidad de votar a favor

In [9]:
from scipy.special import betaln, comb
import math
from scipy.special import gammaln

In [11]:

# Parámetros
n = 3030      # número total de votantes
k = 1515      # número de votos a favor (empate)
p = 0.5       # probabilidad de votar a favor

# Calcular la probabilidad exacta de empate
prob_empate = binom.pmf(k, n, p)

print(f"La probabilidad de un empate exacto es: {prob_empate:.6f} (~{prob_empate*100:.4f}%)")
print("Lo cual no es taaaan raro, dado que la binomial alcanza su máximo en k=n/2, y la probabilidad de empate es la más alta en ese punto.")


La probabilidad de un empate exacto es: 0.014494 (~1.4494%)
Lo cual no es taaaan raro, dado que la binomial alcanza su máximo en k=n/2, y la probabilidad de empate es la más alta en ese punto.


In [ ]:
# Esta versión genera 
# OverflowError: int too large to convert to float
n = 3030
k = 1515

# Calcular coeficiente binomial
coef_binomial = comb(n, k, exact=True)

# Calcular la función beta B(k+1, n-k+1)
beta_val = beta(k + 1, n - k + 1)

# Probabilidad bayesiana (beta-binomial con Beta(1,1) previa)
prob_bayesiana = coef_binomial * beta_val

print(f"Probabilidad de empate bajo enfoque bayesiano: {prob_bayesiana:.8f} (~{prob_bayesiana*100:.6f}%)")


OverflowError: int too large to convert to float

In [10]:
n = 3030
k = 1515

# Calcular logaritmo del coeficiente binomial usando gammaln para precisión numérica
# log comb(n,k) = gammaln(n+1) - gammaln(k+1) - gammaln(n-k+1)

log_coef_binomial = gammaln(n + 1) - gammaln(k + 1) - gammaln(n - k + 1)

# Calcular logaritmo de la función beta
log_beta_val = betaln(k + 1, n - k + 1)

# Sumar logaritmos
log_prob_bayesiana = log_coef_binomial + log_beta_val

# Exponenciar para obtener la probabilidad
prob_bayesiana = math.exp(log_prob_bayesiana)

print(f"Probabilidad de empate bajo enfoque bayesiano: {prob_bayesiana:.8e} (~{prob_bayesiana*100:.8f}%)")


Probabilidad de empate bajo enfoque bayesiano: 3.29924117e-04 (~0.03299241%)


## Enfoque Bayesiano

### Modelo:

Los votos a favor \( X \) siguen una distribución binomial condicional a \( p \):

$$
X \mid p \sim \text{Binomial}(n=3030, p)
$$

La probabilidad \( p \) tiene una distribución previa uniforme:

$$
p \sim \text{Beta}(\alpha=1, \beta=1)
$$

### Posterior predictiva:

La probabilidad de observar un empate exacto \( X = 1515 \) sin conocer \( p \) se calcula integrando sobre todos los posibles valores de \( p \):

$$
P(X=1515) = \int_0^1 P(X=1515 \mid p) f(p) \, dp = \int_0^1 \binom{3030}{1515} p^{1515} (1-p)^{1515} \, dp
$$

Dado que $ p \sim \text{Beta}(1,1) $, la integral es la función beta:

$$
P(X=1515) = \binom{3030}{1515} \frac{B(1515+1, 1515+1)}{B(1,1)} = \binom{3030}{1515} B(1516, 1516)
$$

donde $ B(a,b) $ es la función beta.

### Resultado:

Esta probabilidad bajo el modelo beta-binomial es aproximadamente igual a:

$$
P(X=1515) \approx \frac{1}{3031} \approx 0.00033 = 0.033\%
$$

Esto representa la probabilidad de empate considerando que no sabemos nada sobre $ p $ y que todos los valores de $ p $ son igualmente posibles a priori.

### Interpretación:

- Bajo este enfoque bayesiano con una distribución previa uniforme para $ p $, la probabilidad de empate es mucho menor que bajo el modelo frecuentista con $ p = 0.5 $ fijo.
- Sin embargo, si se incorpora información previa o se usa una distribución beta más informativa, la probabilidad puede acercarse al valor frecuentista (~1.44%).

---
